packages/core/src/readers/LlamaParseReader.ts

```diff
@@ -188,48 +187,170 @@ export class LlamaParseReader implements FileReader {
+    const signal = AbortSignal.timeout(this.maxTimeout * 1000);
     let tries = 0;
     while (true) {
       await new Promise((resolve) =>
         setTimeout(resolve, this.checkInterval * 1000),
       );
-      response = await fetch(resultUrl, {
+
+      // Check the job status. If unsuccessful response, checks if maximum timeout has been reached. If reached, throws an error
+      const statusResponse = await fetch(statusUrl, {
         headers,
-        signal: AbortSignal.timeout(this.maxTimeout * 1000),
+        signal,
       });
+      if (!statusResponse.ok) {
+        signal.throwIfAborted();
+        if (this.verbose && tries % 10 === 0) {
+          process.stdout.write(".");
+        }
+        tries++;
+        continue;
+      }
 
-      if (!response.ok) {
-        const end = Date.now();
-        if (end - start > this.maxTimeout * 1000) {
+      // If response is succesful, check status of job. Allowed values "PENDING", "SUCCESS", "ERROR", "CANCELED"
+      const statusJson = await statusResponse.json();
+      const status = statusJson.status;
+      // If job has completed, return the result
+      if (status === "SUCCESS") {
+        const resultResponse = await fetch(resultUrl, {
+          headers,
+          signal,
+        });
+        if (!resultResponse.ok) {
           throw new Error(
-            `Timeout while parsing the file: ${await response.text()}`,
+            `Failed to fetch result: ${await resultResponse.text()}`,
           );
         }
+        return resultResponse.json();
+        // If job is still pending, check if maximum timeout has been reached. If reached, throws an error
+      } else if (status === "PENDING") {
+        signal.throwIfAborted();
         if (this.verbose && tries % 10 === 0) {
           process.stdout.write(".");
         }
         tries++;
-        continue;
+      } else {
+        throw new Error(
+          `Failed to parse the file: ${jobId}, status: ${status}`,
+        );
       }
+    }
+  }
+
+  /**
+   * Loads data from a file and returns an array of Document objects.
+   * To be used with resultType = "text" and "markdown"
+   *
+   * @param {string} file - The path to the file to be loaded.
+   * @return {Promise<Document[]>} A Promise object that resolves to an array of Document objects.
+   */
+  async loadData(file: string): Promise<Document[]> {
+    // Set metadata to contain file_path
+    const metadata = { file_path: file };
 
-      const resultJson = await response.json();
-      return [
-        new Document({
-          text: resultJson[this.resultType],
-          metadata: metadata,
-        }),
-      ];
+    // Creates a job for the file
+    const jobId = await this.createJob(file);
+    if (this.verbose) {
+      console.log(`Started parsing the file under job id ${jobId}`);
+    }
+
+    // Return results as Document objects
+    const resultJson = await this.getJobResult(jobId, this.resultType);
+    return [
```

Comment:
Should this return a list of documents by page?

Reply:
@himself65 I don't think so, as far as I understand in LlamaParse only json mode relies on pages and even than it seems mostly for the image retrieval functionality. Documentation is a bit spare and so is my knowledge of python, so take it with a grain of salt and maybe check it yourself. But from my understanding. I think it's fine to return the full documents and split them up later instead of listing them by page